darshan-util/pydarshan/darshan/experimental/plots/plot_dxt_heatmap.py

```diff
@@ -0,0 +1,425 @@
+
+
+def set_y_axis_ticks_and_labels(jointgrid: Any, n_ylabels: int = 6):
+    """
+    Sets the y-axis tick mark locations and labels.
+
+    Parameters
+    ----------
+    jointgrid: a ``sns.axisgrid.JointGrid`` object.
+
+    n_ylabels: The number of y-axis tick mark labels to create. Default is 6.
+
+    """
+    # retrieve the y-axis tick mark locations and labels
+    yticks = get_y_axis_ticks(ax=jointgrid.ax_joint, n_ylabels=n_ylabels)
+    yticklabels = get_y_axis_tick_labels(ax=jointgrid.ax_joint, n_ylabels=n_ylabels)
+    # set the new y-axis tick locations and labels
+    jointgrid.ax_joint.set_yticks(yticks)
+    jointgrid.ax_joint.set_yticklabels(yticklabels, minor=False)
+
+
+def remove_marginal_graph_ticks_and_labels(marg_x: Any, marg_y: Any):
+    """
+    Removes the frame, tick marks, and tick mark
+    labels for the marginal bar graphs.
+
+    Parameters
+    ----------
+    marg_x : a x-axis marginal bar graph object.
+
+    marg_y : a y-axis marginal bar graph object.
+    """
+    for marg in [marg_x, marg_y]:
+        # turn the frame off for both bar graphs
+        marg.axis("off")
+        # remove all tick mark labels for both bar graphs
+        marg.tick_params(
+            axis="x", bottom=False, labelbottom=False, top=False, labeltop=False
+        )
+        marg.tick_params(
+            axis="y", left=False, labelleft=False, right=False, labelright=False
+        )
+
+
+def adjust_for_colorbar(jointgrid: Any, fig_right: float, cbar_x0: float):
+    """
+    Makes various subplot location adjustments such that
+    a colorbar can fit in the overal figure panel.
+
+    Parameters
+    ----------
+
+    jointgrid: a ``sns.axisgrid.JointGrid`` object.
+
+    fig_right: the location to set for the right side of the heatmap figure.
+
+    cbar_x0: the x-axis location of the colorbar.
+
+    """
+    # adjust the subplot so the x/y tick labels are legible
+    jointgrid.fig.subplots_adjust(
+        left=0.1, bottom=0.15, top=0.9, hspace=0.03, wspace=0.04
+    )
+    # set the location of the right side of the figure
+    jointgrid.fig.subplots_adjust(right=fig_right)
+    # get the positions of the joint and marginal x axes
+    pos_joint_ax = jointgrid.ax_joint.get_position()
+    pos_marg_x_ax = jointgrid.ax_marg_x.get_position()
+    # set the position and dimensions of the joint plot such that it fills
+    # the space as if there was no colorbar
+    jointgrid.ax_joint.set_position(
+        [pos_joint_ax.x0, pos_joint_ax.y0, pos_marg_x_ax.width, pos_joint_ax.height]
+    )
+    # set the position of the colorbar such that it is on the
+    # right side of the horizontal bar graph, and set its dimensions
+    jointgrid.fig.axes[-1].set_position(
+        [cbar_x0, pos_joint_ax.y0, 0.9, pos_joint_ax.height]
+    )
+
+
+def plot_heatmap(
+    log_path: str,
+    mods: Sequence[str] = ["DXT_POSIX"],
+    ops: Sequence[str] = ["read", "write"],
+    xbins: int = 200,
+) -> Any:
```

Comment:
since `plot_heatmap()` is likely to be "public"/used externally to some extent, may be worth adding some more error handling for things that currently cause errors in NumPy itself, like negative/invalid bin counts (see below). Can open an issue for that for now if you want; using `hypothesis` a little on the public function APIs may poke a few holes where we could provide helpful errors I suppose.

```python
from darshan.experimental.plots import plot_dxt_heatmap

log_path = './tests/input/sample-dxt-simple.darshan'

plot_dxt_heatmap.plot_heatmap(log_path=log_path,
                              xbins=-17)

plot_dxt_heatmap.plot_heatmap(log_path=log_path,
                              xbins='junk')
```

Reply:
That's a good point -- I'll open an issue shortly. 

I haven't been too concerned about checking for invalid inputs since whatever framework we go with will likely do most of the input handling for the plotting functions. At least that's what I was anticipating. But with that in mind, it is still likely to help develop the framework things.

Reply:
I think some of the plotting code will be exposed as public for individual use as well.

Reply:
I don't know that the framework will do too much input handling apart from providing sensible defaults most of the time, otherwise if you have 10 plots in the report the command line might get pretty wild specifying stuff for each one.

Also, separating the input handling from each individual function into the framework might get pretty messy (you'd get dissociation between an individual function and its error handling).